Refactoring and modularization of user code migration process

cboulanger opened the pull request:
still unfinished & unfunctional, just showing the current thinking.

cboulanger committed:
First refactoring of migration, still unfunctional

derrell commented on source/class/qx/tool/cli/Utils.js:
```diff
@@ -0,0 +1,105 @@
+   http://qooxdoo.org
+
+   Copyright:
+     2017-202 The authors
```
year invalid

derrell commented on source/class/qx/tool/cli/Utils.js:
```diff
@@ -0,0 +1,105 @@
+        exe.on("close", code => {
+          if (code !== 0) {
+            let message = `Error executing '${cmd} ${args.join(" ")}'. Use --verbose to see what went wrong.`;
+            throw new qx.tool.utils.Utils.UserError(message);
```
Should it throw here, or better to call `reject`?

cboulanger replied:
done.

derrell commented on source/class/qx/tool/migration/IMigration.js:
```diff
@@ -0,0 +1,38 @@
+   http://qooxdoo.org
+
+   Copyright:
+     2017-202 The authors
```
year invalid

derrell commented on source/class/qx/tool/migration/M6_0_0.js:
```diff
@@ -0,0 +1,183 @@
+      let pkg = qx.tool.cli.commands.Package;
+      let cwd = process.cwd();
+      let migrateFiles = [[path.join(cwd, pkg.lockfile.filename), path.join(cwd, pkg.lockfile.legacy_filename)], [path.join(cwd, pkg.cache_dir), path.join(cwd, pkg.legacy_cache_dir)], [path.join(qx.tool.cli.ConfigDb.getDirectory(), pkg.package_cache_name), path.join(qx.tool.cli.ConfigDb.getDirectory(), pkg.legacy_package_cache_name)]];
+      if (qx.tool.migration.Utils.checkFilesToRename(migrateFiles).length) {
```
Some comments in here saying what this code is trying to accomplish would be really useful. It's not obvious.

cboulanger replied:
The migration code has been modularized, with each step in one separate method- the name should tell you what it does: https://github.com/cboulanger/qooxdoo/blob/refactor_migration/source/class/qx/tool/migration/M6_0_0.js

derrell reviewed (commented):
This is WIP, but I took a look through it anyway. Lots going on here.

hkollmann commented:
What's about adding some unit tests in qx.test.tool.cli namespace?

cboulanger commented:
@cboulanger Oh yes certainly. This PR is only 1/3-finished. Just wanted to let you guys know the direction it intends to take.

cboulanger committed:
Fix typo

cboulanger committed:
Revert factoring out of the getUserX() methods

cboulanger committed:
Reject not throw

hkollmann committed:
add rules so that lint runs error free (#10126)

fix compiler warnings
remove packages
more meaningfull log message

hkollmann committed:
seperate compiler from framework tests (#10127)

* seperate compiler from framework tests

* remove gui settings

cboulanger committed:
Small fixes

cboulanger committed:
Merge branch 'master' into refactor_migration

cboulanger committed:
Make compiler run (migration still disabled)

cboulanger committed:
More architectural refactoring, not yet in a functional state

cboulanger committed:
New version of getQxPath

cboulanger committed:
Fixes, removal of block-global-framework param

cboulanger committed:
Merge branch 'master' into refactor_migration

cboulanger committed:
Remove `qx.library` setting documentation & add to changelog

cboulanger committed:
Deprecate qx pkg migrate in favor of qx migrate

cboulanger committed:
Fixes and switch to qx.log logging

cboulanger committed:
Refactor return value of migration methods

cboulanger committed:
Fixes

cboulanger committed:
Fixes

cboulanger committed:
Test runs

cboulanger committed:
Merge branch 'master' into refactor_migration

cboulanger commented:
Ready for review.

hkollmann reviewed (commented):
is  test/cli/compiler/test-migrations/v6.0.0/unmigrated/contrib/.gitkeep needed?

cboulanger commented on bin/tools/utils.js:
```diff
@@ -216,6 +218,379 @@ fs.copyFileSync("bin/build/qx.cmd", "bootstrap/qx.cmd");
+   * @param length {Integer} maximum number of files
+   * @async
+   */
+  rotateUnique: async function(filename, length) {
```
Oh, I just copy & pasted utility methods from  `qx.tool.utils.*` to make them available to the test methods, which do not have qx available. The state of utility methods in `qx.tool` is a real mess, which has to do with its ad-hoc and decentral development. There is a lot of redundancy and overlapping methods, plus duplicate functionality targeting earlier node versions. It needs a thorough clean-up, but that means a lot of work. In particular, we need some classes which can be accessed both by qx and the nodeJS runtime - which would probably work with the "qx.server" package.

The method you point to is probably not needed by the test pipeline and can be removed.

cboulanger replied:
I think `.gitkeep` is needed because git doesn't store empty directories. I need the directoy because it must be renamed by the migration.

johnspackman replied:
@hkollmann if it's not used anywhere then it can be deleted; I recognise it as one of mine, I use it for rotating log files, but it's not very well defined (or sophisticated) so does not need to be in the framework unless its used

hkollmann replied:
@cboulanger : Could you delete it? Inclusive the then not needed test?

cboulanger replied:
I will delete it.

cboulanger replied:
Done.

cboulanger committed:
Remove unneeded utility method

cboulanger commented:
Copying this over from Gitter: What this PR does is to allow granular migrations, i.e. for every version change (of course this mostly applies to breaking ones), a migration file can be added (e.g. `qx.tool.migration.M7_1_3`). Ideally, our workflow would be configured to require that any breaking change MUST contain such a file, so it is guaranteed to provide a migration path. If the user updates the framework, the CLI will warn that migrations exist that have to be applied and tell the user to qx migrate.

The migration classes should extend from `qx.tool.migration.BaseMigration` which contains useful utility methods. Any method starting with "migration" will be run, which makes it easy to organize migration steps easily. Here is an example for the 6.0.0-beta to 6.0.0 migration: https://github.com/cboulanger/qooxdoo/blob/refactor_migration/source/class/qx/tool/migration/M6_0_0.js

The PR also includes:
- the removal of `qx.library` config setting
- change of the way the qooxdoo library is looked up - now the canonical methods are 
   - `qx.tool.config.Utils.getQxPath()` to get the qooxdoo version in the environment (i.e. usually the NPM module, but it will autodetect the repo also) of the current working dir, 
   - `qx.tool.config.Utils.getAppQxPath()` to get the qooxdoo version used by the package/library/application in the current working dir; this can be different from the `getQxPath()` value. It is calculated either by looking up the qooxdo library referenced in the `libraries` key of `compile.json`, which takes precedence, or by looking at the `@qooxdoo/framework` dependency information in the Manifest. 
- allows to use the `this.(debug|info|warn|error)` methods from within server code.

derrell commented on source/class/qx/tool/cli/commands/Command.js:
```diff
@@ -60,353 +57,129 @@ qx.Class.define("qx.tool.cli.commands.Command", {
+        this.debug(`Not checking migration because check is already in progress.`);
+      } catch (e) {
+        // run migration in dry-run mode
+        await fsp.writeFile(semaphore,"");
```
We've got a race condition here. Since we're using a semaphore, we're clearly concerned with two processes doing migrations at the same time... but it's possible that the `stat` runs in one of the processes and finds no error, an operating system task switch occurs and `stat` runs in the other process finding no error, and as each process continues, they each enter `writeFile`. In general, it is better not to do a test to see if you can do something and then do it (which causes the race condition), but rather to just "try to do it" and let it fail if it can't.

In this case, that would be done by getting rid of the `stat` and passing options to `writeFile` to tell it to error if the file exists. This sample code shows the general idea: if /tmp/xxx already exists, this will fail.

```
fsp.writeFile("
  /tmp/xxx", 
  "hello world",
  {
    flag : fs.constants.O_WRONLY | fs.constants.O_CREAT | fs.constants.O_EXCL,
    mode: 0o644
  })
 .then(() => console.log("worked"))
 .catch((e) => console.log("failed: " + e));
```

cboulanger replied:
You are right in principle, but the use case are only for the case that one `qx` command triggers another one before it has finished, and the chance of a race condition here is almost zero because of the time it takes before a `qx` CLI command actually executes. If this was mission-critical, I'd rewrite it, but it is probably not even real problem - I was already thinking to remove it altogether.

derrell commented on source/class/qx/tool/config/Utils.js:
```diff
@@ -0,0 +1,308 @@
+        }
+        // 2. node_modules folders
+        let npmdir = path.join(dir, "node_modules", "@qooxdoo", "framework")
+        if (fs.existsSync(path.join(npmdir, "Manifest.json"))) {
```
consider replacing sync call with async one?

cboulanger replied:
Done. Thanks.

derrell commented on source/class/qx/tool/migration/BaseMigration.js:
```diff
@@ -0,0 +1,225 @@
+    checkFilesToRename(fileList) {
+      let filesToRename = [];
+      for (let [newPath, oldPath] of fileList) {
+        if (!fs.existsSync(newPath) && fs.existsSync(oldPath)) {
```
sync -> async ?

derrell commented on source/class/qx/tool/migration/BaseMigration.js:
```diff
@@ -0,0 +1,225 @@
+    async checkFilesContain(files, text) {
+      files = Array.isArray(files) ? files : [files];
+      for (let file of files) {
+        if ((await fsp.stat(file)).isFile() && (await fsp.readFile(file, "utf8")).includes(text)) {
```
This looks to me like if `fsp.stat().isFile()` returns false, we'll be doing `if false.includes(text)` which will throw. Is that what's intended?

cboulanger replied:
As per the Docs, the method always returns a <Stats> object. https://nodejs.org/api/fs.html#fs_class_fs_stats even if there is no file.

derrell replied:
That's ugly. A function that begins "is" generally returns a boolean. Anyway, not our problem.

derrell commented on source/class/qx/tool/migration/M6_0_0.js:
```diff
@@ -0,0 +1,226 @@
+      let migrationInfo = this.getRunner().createMigrationInfo();
+      let compileJsFilename = path.join(process.cwd(), "compile.js");
+      if (await fs.existsAsync(compileJsFilename)) {
+        let data = await fs.readFileAsync(compileJsFilename, "utf8");
```
`existsAsync` looks unnecessary. Is there a reason not to simply try to read the file, and ignore the ENOEXIST error? (This one is likely not a race condition; just seems superfluous.)

cboulanger replied:
Sure, try/catch would work here just the same. This is just habit. But there is a drawback - if you just `try` the readFileAsync in order not to catch errors from the other code inside the `if` clause, you have to set some flag insde the try:
```javascript
let flag = false;
try { 
  let data = await fs.readFileAsync(compileJsFilename, "utf8");
  flag = true;
} catch(e){}
if (flag) { 
// do the other code that was inside the if clause
} 
 ```

I am not sure this is really superior.

derrell replied:
Wouldn't `finally` solve that?

Not important. Was just a thought.

cboulanger replied:
You are right! That's exactly what `finally` is for. I forgot about it.

cboulanger replied:
(I think when I first learned javascript, it didn't have finally)

hkollmann committed:
Merge branch 'master' into refactor_migration

cboulanger commented on bin/tools/utils.js:
```diff
@@ -5,19 +5,21 @@ const child_process = require("child_process");
 //var fsPromises = require("fs").promises;
 // node 8 compatibility
 const {promisify} = require('util');
+const stat = promisify(fs.stat);
```
The whole file is a basket full of random utility functions with different dependencies, so this is just as it is toto make it work. 

It would be a waste of time to clean up this file because it needs to be replaced by a fully rewritten static class that is included in the "qooxdoo" (qx.server) npm package and this way can be accessed by nodeJS. As I have written elsewhere, the whole utility method business requires a rewrite, but this is outside the scope of this PR.

cboulanger commented on bin/tools/utils.js:
```diff
@@ -216,6 +218,355 @@ fs.copyFileSync("bin/build/qx.cmd", "bootstrap/qx.cmd");
+          p = Promise.resolve();
+        }
+        return p.then(() => fsPromises.readdir(from)
+          .then(files => Promise.all(files.map(file => t.sync(path.join(from, file), path.join(to, file), filter)))));
```
This function is just copy & pasted from another `qx.tool.utils` utility function which is not accessible to the NodeJS process. Same problem as before, so please just ignore it for the moment.

cboulanger committed:
Make sync method async

cboulanger committed:
Make sync method async

hkollmann commented on test/cli/compiler/test-migrations.js:
```diff
@@ -0,0 +1,29 @@
+    await testUtils.deleteRecursive(migratedDir);
+    await testUtils.sync(unmigratedDir, migratedDir);
+    console.log("\n# upgrade alert\n");
+    await testUtils.runCommand(migratedDir, qx, "clean");
```
Why do not use runcompiler here?

cboulanger replied:
What would be the equivalent commands?

hkollmann commented:
Sorry runCompiler only runs compile process.
But there is a helper in utils to fetch qx - so there is no need to do it yourself

Von meinem iPad gesendet

> Am 19.02.2021 um 20:18 schrieb Christian Boulanger <notifications@github.com>:
> 
> ﻿
> @cboulanger commented on this pull request.
> 
> In test/cli/compiler/test-migrations.js:
> 
> > +const path = require("path");
> +const process = require("process");
> +const testUtils = require("../../../bin/tools/utils");
> +const fsp = require("fs").promises;
> +const qx = path.resolve(__dirname, "..", "..", "..", "bin", "source", "qx");
> +const digest = require("util").promisify(require("dirsum").digest);
> +
> +test("Test migrations", async assert => {
> +  try {
> +    const baseDir = path.join(__dirname, "test-migrations", "v6.0.0");
> +    const unmigratedDir = path.join(baseDir, "unmigrated");
> +    const migratedDir = path.join(baseDir, "migrated");
> +    await testUtils.deleteRecursive(migratedDir);
> +    await testUtils.sync(unmigratedDir, migratedDir);
> +    console.log("\n# upgrade alert\n");
> +    await testUtils.runCommand(migratedDir, qx, "clean");
> What would be the equivalent commands?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

hkollmann commented on test/cli/compiler/test-migrations.js:
```diff
@@ -0,0 +1,29 @@
+const process = require("process");
+const testUtils = require("../../../bin/tools/utils");
+const fsp = require("fs").promises;
+const qx = path.resolve(__dirname, "..", "..", "..", "bin", "source", "qx");
```
const qx = testUtils.getCompiler()

cboulanger replied:
Strangely, that breaks the tests, because getCompiler() returns the built version of the compiler instead of the source version. However, that means that there's something wrong, since both versions should produce the same output. Let me check.

cboulanger committed:
Add docs

cboulanger committed:
Update docs

cboulanger committed:
Refactor migration info API

cboulanger committed:
Update tests

cboulanger commented:
@hkollmann The problem was that I hadn't recompiled the "build" compiler. Now everything works in source or build mode.

I have changed the API for updating runtime migration information. Instead of passing around an object with that info, I added two methods `markAsApplied()` and `markAsPending()` to the base class. Much simpler and cleaner and less prone to breakage.

Now everything is ready. @hkollmann, once you have approved, you can merge it!

cboulanger committed:
Remove docs for non-existing `qx upgrade`

(there is only "qx pkg upgrade")

cboulanger committed:
Fix documentation
